app/Models/Beatmapset.php

```diff
@@ -726,6 +728,40 @@ public function nominate(User $user)
         return true;
     }
 
+    public function favorite($user)
+    {
+        DB::transaction(function () use ($user) {
+            try {
+                FavoriteBeatmapset::create([
+                    'user_id' => $user->user_id,
+                    'beatmapset_id' => $this->beatmapset_id,
+                ]);
+            } catch (QueryException $e) {
+                if (is_sql_unique_exception($e)) {
+                    return;
+                }
+            }
+
+            $this->favourite_count += 1;
+            $this->save();
+        });
+    }
+
+    public function unfavorite($user)
+    {
+        if (!$this->hasFavorited($user)) {
+            return;
+        }
+
+        DB::transaction(function () use ($user) {
+            $this->favorites()->where('user_id', $user->user_id)
+                ->delete();
+
+            $this->favourite_count -= 1;
```

Comment:
here as well, with `GREATEST(..., 0)` to avoid updating to negative. See `app/Models/Forum/PollOption.php`.